pkg/issuer/acme/issue.go

```diff
@@ -74,38 +78,88 @@ func (a *Acme) obtainCertificate(ctx context.Context, crt *v1alpha1.Certificate)
 		return nil, nil, fmt.Errorf("error getting certificate private key: %s", err.Error())
 	}
 
-	// generate a csr
-	template, err := pki.GenerateCSR(a.issuer, crt)
-	if err != nil {
-		// TODO: this should probably be classed as a permanant failure
-		return nil, nil, err
+	var certSlice [][]byte
+	// StatusReady indicates that the order is ready to be finalized.
+	// Once the order has been finalized, it will transition into the 'valid'
+	// state.
+	// The only way to obtain a certificate from an already 'valid' order is to
+	// call the orders GetCertificate function.
+	// You can see we do this below *iff* certSlice is nil.
+	if order.Status == acmeapi.StatusReady {
+		// generate a csr
+		template, err := pki.GenerateCSR(a.issuer, crt)
+		if err != nil {
+			// TODO: this should probably be classed as a permanant failure
+			return nil, nil, err
+		}
+
+		derBytes, err := pki.EncodeCSR(template, key)
+		if err != nil {
+			crt.UpdateStatusCondition(v1alpha1.CertificateConditionReady, v1alpha1.ConditionFalse, errorIssueError, fmt.Sprintf("Failed to generate certificate request: %v", err), false)
+			return nil, nil, err
+		}
+
+		// obtain a certificate from the acme server
+		certSlice, err = cl.FinalizeOrder(ctx, order.FinalizeURL, derBytes)
+		if err != nil {
+			// this handles an edge case where a certificate ends out with an order
+			// that is in an invalid state.
+			// ideally we would instead call GetCertificate on the ACME client
+			// instead of FinalizeOrder, which would save us creating a new order
+			// just to issue a new certificate.
+			// The underlying ACME client doesn't expose this though yet.
+			if acmeErr, ok := err.(*acmeapi.Error); ok {
+				if acmeErr.StatusCode >= 400 && acmeErr.StatusCode <= 499 {
+					crt.Status.ACMEStatus().Order.URL = ""
+				}
+			}
```

Comment:
I think we might be able to remove this check now. Leaving it in for the time being however, because if the server responds 4xx then we probably should create a new order because we don't have permission/we are sending back requests to it.